homeassistant/components/tplink/light.py

```diff
@@ -72,192 +73,293 @@ def brightness_from_percentage(percent):
+            hue, sat, _ = self.smartbulb.hsv
+            hue_saturation = (hue, sat)
+
+        if self.smartbulb.has_emeter:
+            emeter_params[ATTR_CURRENT_POWER_W] = "{:.1f}".format(
+                self.smartbulb.current_consumption()
+            )
+            daily_statistics = self.smartbulb.get_emeter_daily()
+            monthly_statistics = self.smartbulb.get_emeter_monthly()
+            try:
+                emeter_params[ATTR_DAILY_ENERGY_KWH] = "{:.3f}".format(
+                    daily_statistics[int(time.strftime("%d"))]
+                )
+                emeter_params[ATTR_MONTHLY_ENERGY_KWH] = "{:.3f}".format(
+                    monthly_statistics[int(time.strftime("%m"))]
+                )
+            except KeyError:
+                # device returned no daily/monthly history
+                pass
+
+        return LightState(
+            state=state,
+            brightness=brightness,
+            color_temp=color_temp,
+            hs=hue_saturation,
+            emeter_params=emeter_params,
+        )
+
+    async def async_set_light_state_retry(
+        self, old_light_state: LightState, new_light_state: LightState
+    ) -> bool:
+        """Set the light state with retry."""
+        # Optimistically setting the light state.
+        self._light_state = new_light_state
+
+        # Tell the device to set the states.
+        self._is_setting_light_state = True
+        self._is_available = await self.async_set_light_state(
+            old_light_state, new_light_state
+        ) or await self.async_set_light_state(old_light_state, new_light_state)
+        self._is_setting_light_state = False
+
+        return self._is_available
+
+    async def async_set_light_state(
+        self, old_light_state: LightState, new_light_state: LightState
+    ) -> bool:
+        """Set the light state."""
+        try:
+            # Calling the API with the new state information.
+            if new_light_state.state != old_light_state.state:
+                if new_light_state.state:
+                    self.smartbulb.state = SmartBulb.BULB_STATE_ON
+                else:
+                    self.smartbulb.state = SmartBulb.BULB_STATE_OFF
+                    return True
+
+            if new_light_state.color_temp != old_light_state.color_temp:
+                self.smartbulb.color_temp = mired_to_kelvin(new_light_state.color_temp)
+
+            brightness_pct = brightness_to_percentage(new_light_state.brightness)
+            if new_light_state.hs != old_light_state.hs and len(new_light_state.hs) > 1:
+                hue, sat = new_light_state.hs
+                hsv = (int(hue), int(sat), brightness_pct)
+                self.smartbulb.hsv = hsv
+            elif new_light_state.brightness != old_light_state.brightness:
+                self.smartbulb.brightness = brightness_pct
+
+            return True
+        except (SmartDeviceException, OSError) as ex:
```

Comment:
If we're doing sync I/O in this method we can't do that in async context. We're currently in a coroutine, ie in async context. Sync I/O must be scheduled on the executor thread pool with `hass.async_add_executor_job`.